lib/src/validator/git_status.dart

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:path/path.dart' as p;
+
+import '../git.dart' as git;
+import '../log.dart' as log;
+import '../utils.dart';
+import '../validator.dart';
+
+/// A validator that validates that no checked in files are modiofied in git.
+///
+/// Doesn't report on newly added files, as generated files might not be checked
+/// in to git.
+class GitStatusValidator extends Validator {
+  @override
+  Future<void> validate() async {
+    if (package.inGitRepo) {
+      final modfiedFiles = <String>[];
+      try {
+        final reporoot = git.repoRoot(package.dir);
+        if (reporoot == null) {
+          log.fine(
+            'Could not determine the repository root from ${package.dir}.',
+          );
+          // This validation is only a warning.
+          return;
+        }
+        final output = git.runSyncBytes(
+          [
+            'status',
+            '-z', // Machine parsable
+            '--no-renames', // We don't care about renames.
+
+            '--untracked-files=no', // Don't show untracked files.
+          ],
+          workingDir: package.dir,
+        );
+        // Split at \0.
+        var start = 0;
+        for (var i = 0; i < output.length; i++) {
+          if (output[i] != 0) {
+            continue;
+          }
```

Comment:
Perhaps consider splitting this out into a helper library.. `git.parseMachineStatus` or something.

This might be worth having unit tests cover.